### PR TITLE
Specialize EntityDecoder for strict entities

### PR DESCRIPTION
**Pull request by bplommer**

### CODE REVIEW

**Comment by bplommer** [comment]
I thought about moving this method into `Entity`. The thing that put me off was deciding what to name it. `chunkF`?

**Comment by bplommer** [comment]
Or, we could have something like `Stream#compile`. So `entity.compile.toChunk` etc.

**Comment by armanbilge** [comment]
Here's a a related/different idea: what if we have in `Entity`:
```scala
def toStrict[F[_]: Concurrent]: F[Entity.Strict[Byte]]
```
Then, you can grab the chunk directly.

**Comment by bplommer** [comment]
The thing that slightly concerns me about that is that different `Entity` types have different semantics with respect to http (in ways I don't really understand). I think converting to a different kind of `Entity` just to collect its bytes risks muddying the concepts.

File: core/shared/src/main/scala/org/http4s/EntityDecoder.scala [hunk 1]
```diff
@@ -205,7 +205,13 @@ object EntityDecoder {
...
+    DecodeResult.success(
+      m.entity match {
+        case Entity.Empty => Chunk.empty[Byte].pure[F]
+        case Entity.Strict(chunk) => chunk.pure[F]
+        case d: Entity.Default[F] => d.body.compile.to(Chunk)
+      }
+    )
 
   /** Helper method which simply gathers the body into a single ByteVector */
   private def collectByteVector[F[_]: Concurrent](m: Media[F]): DecodeResult[F, ByteVector] =
```
Comment (wjoel) [inline core/shared/src/main/scala/org/http4s/EntityDecoder.scala:217]: Should maybe change this one similarly?

Comment (wjoel) [inline core/shared/src/main/scala/org/http4s/EntityDecoder.scala:217]: Should maybe change this one similarly?

File: core/shared/src/main/scala/org/http4s/EntityDecoder.scala [hunk 2]
```diff
@@ -205,7 +205,13 @@ object EntityDecoder {
 
   /** Helper method which simply gathers the body into a single Chunk */
   def collectBinary[F[_]: Concurrent](m: Media[F]): DecodeResult[F, Chunk[Byte]] =
-    DecodeResult.success(m.body.chunks.compile.toVector.map(bytes => Chunk.concat(bytes)))
+    DecodeResult.success(
+      m.entity match {
+        case Entity.Empty => Chunk.empty[Byte].pure[F]
+        case Entity.Strict(chunk) => chunk.pure[F]
+        case d: Entity.Default[F] => d.body.compile.to(Chunk)
```
Comment (rossabaker) [inline core/shared/src/main/scala/org/http4s/EntityDecoder.scala:212]: Is this more efficient than what's there?  This line might be a good improvement for a backport.

**Comment by rossabaker** [comment]
I might like `entity.compile.toChunk`, but what other compile ops are there to justify the `compile`?  There's `toByteVector`.

**Comment by bplommer** [comment]
Superseded.

**Pull request closed without merging**